File: src/ldap.js
import ldap from "ldapjs";

export default class LDAP {
	#client = null;
	#basedn = null;
	#peopledn = null;
	#groupsdn = null;

	constructor (url, basedn) {
		const opts = {
			url
		};
		this.#client = new LDAPJS_CLIENT_ASYNC_WRAPPER(opts);
		this.#basedn = basedn;
		this.#peopledn = `ou=people,${basedn}`;
		this.#groupsdn = `ou=groups,${basedn}`;
	}

	async bindUser (uid, password) {
		return await this.#client.bind(`uid=${uid},${this.#peopledn}`, password);
	}

	async getAllUsers () {
		const result = await this.#client.search(this.#peopledn, {
			scope: "one"
		});
		result.users = result.entries;
		return result;
	}

	async addUser (uid, attrs) {
		const userDN = `uid=${uid},${this.#peopledn}`;
		if (!attrs.cn || !attrs.sn || !attrs.userPassword) {
			return {
				ok: false,
				error: {
					code: 100,
					name: "UndefinedAttributeValueError",
					message: "Undefined Attribute Value"
				}
			};
		}
		const entry = {
			objectClass: "inetOrgPerson",
			cn: attrs.cn,
			sn: attrs.sn,
			uid,
			userPassword: attrs.userPassword
		};
		return await this.#client.add(userDN, entry);
	}

	async getUser (uid) {
		const result = await this.#client.search(`uid=${uid},${this.#peopledn}`, {});
		result.user = result.entries[0]; // assume there should only be 1 entry
		return result;
	}

	async modUser (uid, newAttrs) {
		const logger = new LDAP_MULTIOP_LOGGER(`modify ${uid}`);
		for (const attr of ["cn", "sn", "userPassword"]) {
			if (attr in newAttrs && newAttrs[attr]) { // attr should exist and not be undefined or null
				const change = new ldap.Change({
					operation: "replace",
					modification: {
						type: attr,
						values: [newAttrs[attr]]
					}
				});
				await this.#client.modify(`uid=${uid},${this.#peopledn}`, change, logger);
			}
		}
		return logger;
	}

	async delUser (uid) {
		const logger = new LDAP_MULTIOP_LOGGER(`del ${uid}`);
		const userDN = `uid=${uid},${this.#peopledn}`;
		await this.#client.del(userDN, logger);
		const groups = await this.#client.search(this.#groupsdn, {
			scope: "one",
			filter: `(member=uid=${uid},${this.#peopledn})`
		}, logger);
		if (!logger.ok) {
			return logger;
		}
		for (const element of groups.entries) {
			const change = {
				operation: "delete",
				modification: {
					type: "member",
					values: [`uid=${uid},${this.#peopledn}`]
				}
			};
			await this.#client.modify(element.dn, change, logger);
		}
		return logger;
	}

	async getAllGroups () {
		const result = await this.#client.search(this.#groupsdn, {
			scope: "one"
		});
		result.groups = result.entries;
		return result;
	}

	async addGroup (gid) {
		const groupDN = `cn=${gid},${this.#groupsdn}`;
		const entry = {
			objectClass: "groupOfNames",
			member: "",
			cn: gid
		};
		return await this.#client.add(groupDN, entry);
	}

	async getGroup (gid) {
		const result = await this.#client.search(`cn=${gid},${this.#groupsdn}`, {});
		result.group = result.entries[0]; // assume there should only be 1 entry
		return result;
	}

	async delGroup (gid) {
		const groupDN = `cn=${gid},${this.#groupsdn}`;
		return await this.#client.del(groupDN);
	}

	async addUserToGroup (uid, gid) {
		// add the user
		const change = new ldap.Change({
			operation: "add",
			modification: {
				type: "member",
				values: [`uid=${uid},${this.#peopledn}`]
			}
		});
		return await this.#client.modify(`cn=${gid},${this.#groupsdn}`, change);
	}

	async delUserFromGroup (uid, gid) {
		const change = new ldap.Change({
			operation: "delete",
			modification: {
				type: "member",
				values: [`uid=${uid},${this.#peopledn}`]
			}
		});
		return await this.#client.modify(`cn=${gid},${this.#groupsdn}`, change);
	}
}

class LDAP_MULTIOP_LOGGER {
	op = null;
	ok = true;
	error = [];
	subops = [];
	constructor (op) {
		this.op = op;
	}

	push (op) {
		if (!op.ok) {
			this.ok = false;
			this.error.push(op.error);
		}
		this.subops.push(op);
	}
}

class LDAPJS_CLIENT_ASYNC_WRAPPER {
	#client = null;
	constructor (opts) {
		this.#client = ldap.createClient(opts);
		this.#client.on("error", (err) => {
			console.error(`An error occured:\n${err}`);
		});
		this.#client.on("connectError", (err) => {
			console.error(`Unable to connect to ${opts.url}:\n${err}`);
		});
	}

	#parseError (err) {
		if (err) {
			return { code: err.code, name: err.name, message: err.message };
		}
		else {
			return null;
		}
	}

	bind (dn, password, logger = null) {
		return new Promise((resolve) => {
			this.#client.bind(dn, password, (err) => {
				const result = { op: `bind ${dn}`, ok: err === null, error: this.#parseError(err) };
				if (logger) {
					logger.push(result);
				}
				resolve(result);
			});
		});
	}

	add (dn, entry, logger = null) {
		return new Promise((resolve) => {
			this.#client.add(dn, entry, (err) => {
				const result = { op: `add ${dn}`, ok: err === null, error: this.#parseError(err) };
				if (logger) {
					logger.push(result);
				}
				resolve(result);
			});
		});
	}

	search (base, options, logger = null) {
		return new Promise((resolve) => {
			this.#client.search(base, options, (err, res) => {
				if (err) {
					return resolve({ op: `search ${base}`, ok: false, error: err });
				}
				const result = { op: `search ${base}`, ok: false, error: null, entries: [] };
				res.on("searchRequest", (searchRequest) => { });
				res.on("searchEntry", (entry) => {
					const attributes = {};
					for (const element of entry.pojo.attributes) {
						attributes[element.type] = element.values;
					}
					result.entries.push({ dn: entry.pojo.objectName, attributes });
				});
				res.on("searchReference", (referral) => { });
				res.on("error", (err) => {
					result.ok = false;
					result.error = this.#parseError(err);
					if (logger) {
						logger.push(result);
					}
					resolve(result);
				});
				res.on("end", (res) => {
					result.ok = true;
					result.error = null;
					if (logger) {
						logger.push(result);
					}
					resolve(result);
				});
			});
		});
	}

	modify (name, changes, logger = null) {
		return new Promise((resolve) => {
			this.#client.modify(name, changes, (err) => {
				const result = { op: `modify ${name} ${changes.operation} ${changes.modification.type}`, ok: err === null, error: this.#parseError(err) };
				if (logger) {
					logger.push(result);
				}
				resolve(result);
			});
		});
	}

	del (dn, logger = null) {
		return new Promise((resolve) => {
			this.#client.del(dn, (err) => {
				const result = { op: `del ${dn}`, ok: err === null, error: this.#parseError(err) };
				if (logger) {
					logger.push(result);
				}
				resolve(result);
			});
		});
	}
}

File: app/ldap.go
package app

import (
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/go-ldap/ldap/v3"
)

// LDAPClient wrapper struct containing the connection, baseDN, peopleDN, and groupsDN
type LDAPClient struct {
	client   *ldap.Conn
	basedn   string
	peopledn string
	groupsdn string
}

// returns a new LDAPClient from the config
func NewLDAPClient(config Config) (*LDAPClient, error) {
	LDAPConn, err := ldap.DialURL(config.LdapURL)
	return &LDAPClient{
		client:   LDAPConn,
		basedn:   config.BaseDN,
		peopledn: "ou=people," + config.BaseDN,
		groupsdn: "ou=groups," + config.BaseDN,
	}, err
}

// bind a user using username and password to the LDAPClient
func (l LDAPClient) BindUser(username string, password string) error {
	userdn := fmt.Sprintf("uid=%s,%s", username, l.peopledn)
	return l.client.Bind(userdn, password)
}

func (l LDAPClient) GetAllUsers() (int, gin.H) {
	searchRequest := ldap.NewSearchRequest(
		l.peopledn, // The base dn to search
		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
		"(&(objectClass=inetOrgPerson))",                      // The filter to apply
		[]string{"dn", "cn", "sn", "mail", "uid", "memberOf"}, // A list attributes to retrieve
		nil,
	)

	searchResponse, err := l.client.Search(searchRequest) // perform search
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	var results = []gin.H{} // create list of results

	for _, entry := range searchResponse.Entries { // for each result,
		user := LDAPEntryToLDAPUser(entry)
		results = append(results, LDAPUserToGin(user))
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
		"users": results,
	}
}

func (l LDAPClient) GetUser(uid string) (int, gin.H) {
	searchRequest := ldap.NewSearchRequest( //  setup search for user by uid
		fmt.Sprintf("uid=%s,%s", uid, l.peopledn), // The base dn to search
		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
		"(&(objectClass=inetOrgPerson))",                      // The filter to apply
		[]string{"dn", "cn", "sn", "mail", "uid", "memberOf"}, // A list attributes to retrieve
		nil,
	)

	searchResponse, err := l.client.Search(searchRequest) // perform search
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	entry := searchResponse.Entries[0]

	user := LDAPEntryToLDAPUser(entry)
	result := LDAPUserToGin(user)

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
		"user":  result,
	}
}

func (l LDAPClient) AddUser(uid string, user UserRequired) (int, gin.H) {
	if user.CN == "" || user.SN == "" || user.UserPassword == "" || user.Mail == "" {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": "Missing one of required fields: cn, sn, mail, userpassword",
		}
	}

	addRequest := ldap.NewAddRequest(
		fmt.Sprintf("uid=%s,%s", uid, l.peopledn), // DN
		nil, // controls
	)
	addRequest.Attribute("sn", []string{user.SN})
	addRequest.Attribute("cn", []string{user.CN})
	addRequest.Attribute("mail", []string{user.Mail})
	addRequest.Attribute("userPassword", []string{user.UserPassword})
	addRequest.Attribute("objectClass", []string{"inetOrgPerson"})

	err := l.client.Add(addRequest)
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) ModUser(uid string, user UserOptional) (int, gin.H) {
	if user.CN == "" && user.SN == "" && user.UserPassword == "" && user.Mail == "" {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": "Requires one of fields: cn, sn, mail, userpassword",
		}
	}

	modifyRequest := ldap.NewModifyRequest(
		fmt.Sprintf("uid=%s,%s", uid, l.peopledn),
		nil,
	)
	if user.CN != "" {
		modifyRequest.Replace("cn", []string{user.CN})
	}
	if user.SN != "" {
		modifyRequest.Replace("sn", []string{user.SN})
	}
	if user.Mail != "" {
		modifyRequest.Replace("mail", []string{user.Mail})
	}
	if user.UserPassword != "" {
		modifyRequest.Replace("userPassword", []string{user.UserPassword})
	}

	err := l.client.Modify(modifyRequest)
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) DelUser(uid string) (int, gin.H) {
	userDN := fmt.Sprintf("uid=%s,%s", uid, l.peopledn)

	// assumes that olcMemberOfRefint=true updates member attributes of referenced groups

	deleteUserRequest := ldap.NewDelRequest( // setup delete request
		userDN,
		nil,
	)

	err := l.client.Del(deleteUserRequest) // delete user
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) GetAllGroups() (int, gin.H) {
	searchRequest := ldap.NewSearchRequest(
		l.groupsdn, // The base dn to search
		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
		"(&(objectClass=groupOfNames))", // The filter to apply
		[]string{"cn", "member"},        // A list attributes to retrieve
		nil,
	)

	searchResponse, err := l.client.Search(searchRequest) // perform search
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	var results = []gin.H{} // create list of results

	for _, entry := range searchResponse.Entries { // for each result,
		group := LDAPEntryToLDAPGroup(entry)
		results = append(results, LDAPGroupToGin(group))
	}

	return http.StatusOK, gin.H{
		"ok":     true,
		"error":  nil,
		"groups": results,
	}
}

func (l LDAPClient) GetGroup(gid string) (int, gin.H) {
	searchRequest := ldap.NewSearchRequest( //  setup search for user by uid
		fmt.Sprintf("cn=%s,%s", gid, l.groupsdn), // The base dn to search
		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
		"(&(objectClass=groupOfNames))", // The filter to apply
		[]string{"cn", "member"},        // A list attributes to retrieve
		nil,
	)

	searchResponse, err := l.client.Search(searchRequest) // perform search
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	entry := searchResponse.Entries[0]
	group := LDAPEntryToLDAPGroup(entry)
	result := LDAPGroupToGin(group)

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
		"group": result,
	}
}

func (l LDAPClient) AddGroup(gid string, group Group) (int, gin.H) {
	addRequest := ldap.NewAddRequest(
		fmt.Sprintf("cn=%s,%s", gid, l.groupsdn), // DN
		nil,                                      // controls
	)
	addRequest.Attribute("cn", []string{gid})
	addRequest.Attribute("member", []string{""})
	addRequest.Attribute("objectClass", []string{"groupOfNames"})

	err := l.client.Add(addRequest)
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) ModGroup(gid string, group Group) (int, gin.H) {
	return 200, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) DelGroup(gid string) (int, gin.H) {
	groupDN := fmt.Sprintf("cn=%s,%s", gid, l.groupsdn)

	// assumes that memberOf overlay will automatically update referenced memberOf attributes

	deleteGroupRequest := ldap.NewDelRequest( // setup delete request
		groupDN,
		nil,
	)

	err := l.client.Del(deleteGroupRequest) // delete group
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) AddUserToGroup(uid string, gid string) (int, gin.H) {
	userDN := fmt.Sprintf("uid=%s,%s", uid, l.peopledn)
	groupDN := fmt.Sprintf("cn=%s,%s", gid, l.groupsdn)

	modifyRequest := ldap.NewModifyRequest( // modify group member value
		groupDN,
		nil,
	)

	modifyRequest.Add("member", []string{userDN}) // add user to group member attribute

	err := l.client.Modify(modifyRequest) // modify group
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}

func (l LDAPClient) DelUserFromGroup(uid string, gid string) (int, gin.H) {
	userDN := fmt.Sprintf("uid=%s,%s", uid, l.peopledn)
	groupDN := fmt.Sprintf("cn=%s,%s", gid, l.groupsdn)

	modifyRequest := ldap.NewModifyRequest( // modify group member value
		groupDN,
		nil,
	)

	modifyRequest.Delete("member", []string{userDN}) // remove user from group member attribute

	err := l.client.Modify(modifyRequest) // modify group
	if err != nil {
		return http.StatusBadRequest, gin.H{
			"ok":    false,
			"error": err,
		}
	}

	return http.StatusOK, gin.H{
		"ok":    true,
		"error": nil,
	}
}
